File: methods/survival/templates/python/time_dependent_auc.py
#!/usr/bin/env python3
"""Template: time-dependent (cumulative/dynamic) AUC with IPCW (scikit-survival).

Pass a RISK score (higher = earlier event). For a protective marker, use the
Cox linear predictor, not the raw expression. Apparent AUC on the same data
used to fit the Cox model is optimistic; nest this inside CV or apply it to
a locked external cohort.
"""

from __future__ import annotations

import argparse
import sys

import numpy as np
import pandas as pd
from lifelines import CoxPHFitter
from sksurv.metrics import cumulative_dynamic_auc
from sksurv.util import Surv


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--input", required=True)
    ap.add_argument("--time-col", default="time")
    ap.add_argument("--event-col", default="event")
    ap.add_argument("--marker-col", default="marker")
    ap.add_argument("--times", default="3,6,12", help="evaluation times in the same unit as time")
    args = ap.parse_args()

    df = pd.read_csv(args.input)[[args.time_col, args.event_col, args.marker_col]].dropna()
    y = Surv.from_arrays(event=df[args.event_col].astype(bool), time=df[args.time_col])
    cph = CoxPHFitter().fit(df, args.time_col, args.event_col)
    risk = cph.predict_partial_hazard(df).to_numpy()
    last_event = float(df.loc[df[args.event_col] == 1, args.time_col].max())
    times = [float(t) for t in args.times.split(",") if float(t) < last_event]
    if not times:
        sys.exit(f"no evaluation time is earlier than the last event ({last_event})")
    auc, mean_auc = cumulative_dynamic_auc(y, y, risk, np.array(times))
    print(f"Cox log-HR for {args.marker_col} = {cph.summary.loc[args.marker_col, 'coef']:.3f}")
    for t, a in zip(times, np.atleast_1d(auc)):
        n_evt = int(((df[args.time_col] <= t) & (df[args.event_col] == 1)).sum())
        n_risk = int((df[args.time_col] >= t).sum())
        print(f"AUC({t:g}) = {a:.3f}  events_by={n_evt}  still_at_risk={n_risk}")
    print(f"mean AUC over {times} = {mean_auc:.3f} (apparent; not externally validated)")


if __name__ == "__main__":
    main()

File: methods/survival/templates/R/time_dependent_auc.R
#!/usr/bin/env Rscript
# Template: time-dependent (cumulative/dynamic) AUC with IPCW (timeROC).
# Pass a RISK score. For a protective marker use the Cox linear predictor.

suppressPackageStartupMessages({
  library(survival)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: time_dependent_auc.R input.csv [time_col event_col marker_col times_csv]")
}
infile <- args[[1]]
time_col <- if (length(args) >= 2) args[[2]] else "time"
event_col <- if (length(args) >= 3) args[[3]] else "event"
marker_col <- if (length(args) >= 4) args[[4]] else "marker"
times <- if (length(args) >= 5) {
  as.numeric(strsplit(args[[5]], ",", fixed = TRUE)[[1]])
} else {
  c(3, 6, 12)
}

df <- read.csv(infile, stringsAsFactors = FALSE)
df <- df[complete.cases(df[, c(time_col, event_col, marker_col)]), ]
fit <- coxph(Surv(df[[time_col]], df[[event_col]]) ~ df[[marker_col]], ties = "efron")
risk <- as.numeric(predict(fit, type = "lp"))
last_event <- max(df[[time_col]][df[[event_col]] == 1])
times <- times[times < last_event]
if (!length(times)) {
  stop(sprintf("no evaluation time is earlier than the last event (%.2f)", last_event))
}

if (!requireNamespace("timeROC", quietly = TRUE)) {
  stop("install.packages('timeROC') is required for this template")
}

tr <- timeROC::timeROC(
  T = df[[time_col]],
  delta = df[[event_col]],
  marker = risk,
  cause = 1,
  times = times,
  iid = TRUE
)
print(tr)
cat(sprintf("Cox log-HR = %.3f\n", unname(coef(fit))))
cat("Apparent AUC; nest inside CV or apply to a locked external cohort.\n")
